### PR TITLE
Add MemoryMarshal.GetArrayDataReference(Array)

### LINKED ISSUE #35528: Proposal: MemoryMarshal.GetArrayDataReference<T>(T[,]) overload
__Edit by @GrabYourPitchforks__: See alternative API proposal at https://github.com/dotnet/runtime/issues/35528#issuecomment-706456452.

-----

## Overview

This issue is about adding a new API to `System.Runtime.InteropServices.MemoryMarshal`:

```csharp
namespace System.Runtime.InteropServices
{
    public static partial class MemoryMarshal
    {
        public static ref T GetArrayDataReference<T>(T[,] array);
    }
}
```

This would just be an overload of the existing `MemoryMarshal.GetArrayDataReference<T>(T[])` API ([introduced here](https://github.com/dotnet/runtime/issues/29003)). Additional overloads for 3/4/4D arrays could be considered as well.

## Motivation/behavior

Same as `GetArrayDataReference<T>(T[])`, just extended to more array types. As an additional note, this API would ignore custom lower bounds, if present, and always just return a reference to the first array element, if the array is not empty, no matter its expected index. This API is expected to offer the same behavior as the overload for SZ arrays, and if custom lower bounds are present it's up to individual developers to track them correctly after retrieving the reference.

This API would also be particularly useful considering the fact that the standard index accessor for ND arrays is much slower than the one for SZ arrays (related issue on this point: https://github.com/dotnet/runtime/issues/35293).

cc. @tannergooding

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/35528.

This PR has a few distinct components:

- Introduce the overload `Array.GetArrayDataReference(Array array)`. I opted to name the parameter _array_ instead of _arr_ for local consistency with the existing overload.
- Remove the existing `GetRawSzArrayData` and `GetRawArrayData` internal methods and migrate their callers to the new public APIs.
- In coreclr, define the metasigs for both overloads so that we can intrinsicify the new API in the future without creating ambiguity within the existing _jitinterface.cpp_ helpers.
- In mono, intrinsicify and add interp support for both overloads. (Unlike coreclr, mono uses a consistent array layout across both szarray and mdarray, which made this possible.)
- Shore up the unit tests to cover both overloads.

/cc @buyaa-n, I thought this might interest you because it touches _metasig.h_, which you and I discussed a few weeks back. In this case, I'm introducing two signatures to _metasig.h_: one which takes a `T[]` and returns a `ref T`, and one which takes an `Array` and returns a `ref byte`. Note the use of `DEFINE_METASIG_T` instead of `DEFINE_METASIG` for the `Array`-consuming method. This is because we use the syntax `C(ARRAY)`, which specifically refers to the `System.Array` class, and per comments at the very top of file, you need to use the *\_T* version of the macro if you want to refer to specific classes. These new definitions are then used in _corelib.h_.

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.CoreCLR.cs [hunk 1]
```diff
@@ -20,7 +20,29 @@ public static partial class MemoryMarshal
...
+        /// would have been stored. Such a reference may be used for pinning but must never be dereferenced.
+        /// </summary>
+        /// <exception cref="NullReferenceException"><paramref name="array"/> is <see langword="null"/>.</exception>
+        /// <remarks>
+        /// The caller must manually reinterpret the returned <em>ref byte</em> as a ref to the array's underlying elemental type,
+        /// perhaps utilizing an API such as <em>System.Runtime.CompilerServices.Unsafe.As</em> to assist with the reinterpretation.
+        /// This technique does not perform array variance checks. The caller must manually perform any array variance checks
+        /// if the caller wishes to write to the returned reference.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.CoreCLR.cs:38]:
```suggestion
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        [NonVersionable]
```

**Review by jkotas — APPROVED** [review]
CoreCLR part LGTM. Thanks!

**Comment by vargaz** [comment]
The mono changes look ok.

**Review by stephentoub — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
@jkotas I thought I saw feedback from you earlier that I should add `[NonVersionable]` annotations, but looks like the feedback was deleted? Do you still want those annotations added, or did I imagine this?

**Comment by jkotas** [comment]
> @jkotas I thought I saw feedback from you earlier that I should add [NonVersionable] annotations, but looks like the feedback was deleted? 

Sorry for the confusion. I have deleted that feedback. `GetArrayDataReference(Array array)` cannot not be marked as `[NonVersionable]` since it it depends on MethodTable fields that are not guaranteed to be stable.

File: src/mono/mono/mini/intrinsics.c [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include <mono/utils/mono-memory-model.h>
 
 static GENERATE_GET_CLASS_WITH_CACHE (runtime_helpers, "System.Runtime.CompilerServices", "RuntimeHelpers")
+static GENERATE_GET_CLASS_WITH_CACHE (memory_marshal, "System.Runtime.InteropServices", "MemoryMarshal")
```
Comment (marek-safar) [inline src/mono/mono/mini/intrinsics.c:26]: Should use `GENERATE_TRY_GET_CLASS_WITH_CACHE` as the type is optional

File: src/mono/mono/mini/intrinsics.c [hunk 3]
```diff
@@ -944,6 +940,14 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
 			return ins;
 		} else
 			return NULL;
+	} else if (cmethod->klass == memory_marshal_class) {
```
Comment (marek-safar) [inline src/mono/mono/mini/intrinsics.c:943]: and `mono_class_try_get_XXXX` here

**Comment by GrabYourPitchforks** [comment]
Latest iteration:

- Address Marek's PR feedback
- Improve unit test coverage by exercising tests for arrays of varying rank

**Review by marek-safar — APPROVED** [review]

**Pull request merged**